op-service/txmgr/txmgr_test.go

```diff
@@ -1003,3 +1003,89 @@ func TestNonceReset(t *testing.T) {
 	// internal nonce tracking should be reset every 3rd tx
 	require.Equal(t, []uint64{0, 0, 1, 2, 0, 1, 2, 0}, nonces)
 }
+
+// TestClose ensures that the tx manager will refuse new work and cancel any in progress
+func TestClose(t *testing.T) {
+	conf := configWithNumConfs(1)
+	conf.SafeAbortNonceTooLowCount = 100
+	h := newTestHarnessWithConfig(t, conf)
+
+	// sendTx will fail until it is called a retry-number of times
+	called := 0
+	const retries = 4
+	sendTx := func(ctx context.Context, tx *types.Transaction) (err error) {
+		called += 1
+		if called%retries == 0 {
+			txHash := tx.Hash()
+			h.backend.mine(&txHash, tx.GasFeeCap())
+		} else {
+			time.Sleep(10 * time.Millisecond)
+			err = core.ErrNonceTooLow
+		}
+		return
+	}
+	h.backend.setTxSender(sendTx)
+
+	// demonstrate that a tx is sent, even when it must retry repeatedly
+	ctx := context.Background()
+	_, err := h.mgr.Send(ctx, TxCandidate{
+		To: &common.Address{},
+	})
+	require.NoError(t, err)
+	require.Equal(t, retries, called)
+	called = 0
+
+	// rig the txManager to close during the retrying of the tx
+	// this is technically a race condition in the test, if the go func is not scheduled before the tx is retried fully
+	go func() {
+		time.Sleep(15 * time.Millisecond)
+		h.mgr.Close()
+	}()
+	// demonstrate that a tx will cancel if it is in progress when the manager is closed
+	_, err = h.mgr.Send(ctx, TxCandidate{
+		To: &common.Address{},
+	})
+	require.ErrorIs(t, ErrClosed, err)
+	// confirm that the tx was canceled before it retried to completion
+	require.Less(t, called, retries)
+	called = 0
+
+	// demonstrate that new calls to Send will also fail when the manager is closed
+	_, err = h.mgr.Send(ctx, TxCandidate{
+		To: &common.Address{},
+	})
+	require.ErrorIs(t, ErrClosed, err)
+	// confirm that the tx was canceled before it ever made it to the backend
+	require.Equal(t, 0, called)
+}
+
+// TestCloseWaitingForConfirmation ensures that the tx manager will wait for confirmation of a tx in flight, even when closed
+func TestCloseWaitingForConfirmation(t *testing.T) {
+	// two confirmations required so that we can mine and not yet be fully confirmed
+	conf := configWithNumConfs(2)
+	h := newTestHarnessWithConfig(t, conf)
+
+	sendTx := func(ctx context.Context, tx *types.Transaction) error {
+		txHash := tx.Hash()
+		h.backend.mine(&txHash, tx.GasFeeCap())
+		return nil
+	}
+	h.backend.setTxSender(sendTx)
+
+	// rig the backend to advance confirmations after some time
+	go func() {
+		time.Sleep(100 * time.Millisecond)
+		h.backend.mine(nil, nil)
+	}()
+	// rig the txManager to close between the tx being mined and the tx being confirmed
+	go func() {
+		time.Sleep(50 * time.Millisecond)
+		h.mgr.Close()
+	}()
```

Comment:
I'm wondering if we can get around the sleeps with channel signals, like
```suggestion
    sendDone, closeDone := make(chan struct{}), make(chan struct{})
	sendTx := func(ctx context.Context, tx *types.Transaction) error {
		txHash := tx.Hash()
		h.backend.mine(&txHash, tx.GasFeeCap())
		close(sendDone)
		return nil
	}
	h.backend.setTxSender(sendTx)

	// rig the backend to advance confirmations after some time
	go func() {
		<-closeDone
		h.backend.mine(nil, nil)
	}()
	// rig the txManager to close between the tx being mined and the tx being confirmed
	go func() {
		<-sendDone
		h.mgr.Close()
		close(closeDone)
	}()
```